# Patch to protect gdb from malicious (or corrupt) ELF headers.
# https://bugzilla.redhat.com/bugzilla/show_bug.cgi?id=204845
--- gdb-6.4.90.dfsg/gdb/dwarf2read.c.orig	2006-09-27 16:29:27.000000000 -0700
+++ gdb-6.4.90.dfsg/gdb/dwarf2read.c	2006-09-27 16:29:27.000000000 -0700
@@ -8858,6 +8858,18 @@ decode_locdesc (struct dwarf_block *blk,
 		     dwarf_stack_op_name (op));
 	  return (stack[stacki]);
 	}
+
+	/* Enforce maximum stack depth of 63 to avoid ++stacki writing
+	  outside of the given size. Also enforce minimum >= 0.
+	  -- wad@google.com 14 Aug 2006 */
+	if (stacki >= sizeof(stack)/sizeof(*stack) - 1)
+	  internal_error (__FILE__, __LINE__,
+	                  _("location description stack too deep: %d"),
+	                  stacki);
+	if (stacki < 0)
+	  internal_error (__FILE__, __LINE__,
+	                  _("location description stack too shallow"));
+
     }
   return (stack[stacki]);
 }
--- gdb-6.4.90.dfsg/gdb/dwarfread.c.orig	2006-09-27 16:29:27.000000000 -0700
+++ gdb-6.4.90.dfsg/gdb/dwarfread.c	2006-09-27 16:23:20.000000000 -0700
@@ -2224,6 +2224,18 @@ locval (struct dieinfo *dip)
 	  stacki--;
 	  break;
 	}
+
+    /* Enforce maximum stack depth of 63 to avoid ++stacki writing
+       outside of the given size. Also enforce minimum >= 0.
+       -- wad@google.com 14 Aug 2006 */
+    if (stacki >= sizeof(stack)/sizeof(*stack) - 1)
+      internal_error (__FILE__, __LINE__,
+                      _("location description stack too deep: %d"),
+                      stacki);
+    if (stacki < 0)
+      internal_error (__FILE__, __LINE__,
+                      _("location description stack too shallow"));
+
     }
   return (stack[stacki]);
 }
